In [1]:
import numpy as np
from numpy import linalg as LA
from numpy import array
from numpy import random

import math  
import pandas as pd

from scipy.linalg import svd

from sklearn import preprocessing
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import TransformedTargetRegressor
from sklearn.compose import ColumnTransformer

from matplotlib import pyplot as plt
import seaborn as sns

import time

In [ ]:
# Todo

# EXERCICE 1
# Not done

# EXERCICE 2
# Double check results
# Refactor using functions

# EXERCICE 3
# refactorer pr n'avoir qu'une fonction qui donne iter_u iter_v: plot les 2 sur même graphe?
# Question 8 not found
# Bonus questions not done
# can you show it mathematically:
# https://www.cs.yale.edu/homes/el327/datamining2013aFiles/07_singular_value_decomposition.pdf
# https://www.cs.cmu.edu/~venkatg/teaching/CStheory-infoage/book-chapter-4.pdf

# EXERCICE 5
# split cat and num columns, create 2 pipelines (one for cat and one for num), 
# using a preprocessor like in this article 
# https://medium.com/@oluwabukunmige/implementing-linear-regression-with-sci-kit-learn-b6f87edc3150

# VIF Factor above 10 indicates multicollinearity. cyclinders has highesht VIF factor.
# Ridge regression (+ CV?)
# Lasso regression
# https://medium.com/analytics-vidhya/multiple-linear-regression-7727a012ff93
# https://medium.com/@powusu381/multiple-regression-in-python-using-scikit-learn-predicting-the-miles-per-gallon-mpg-of-cars-4c8e512234be

Exercice 1

Exercice 2

3) Choose three non-Gaussian probability distributions, with mean 0 and variance 2, and
write a function that that takes as input n, p and the distribution name, and creates a
matrix X P R nˆp with entries generated (i.i.d.) according to this distribution.

The 3 non-Gaussian probability distribution with mean 0 and var 2 we choose are <br>
A uniform distribution over [-sqrt(6), sqrt(6)] (with sqrt(6) ~ 2.45) <br>
A Laplace distribution with mean mu=0 and b=1 <br>
A Logistic distribution with mean mu=0 and s=sqrt(6)/pi

4) Display on one single graph the singular values of X for n “ 1000, and p “
200, 500, 1000, 2000 for the three distributions chosen

In [ ]:
def generator(n, p, distribution):
    if distribution == "uniform":
        return np.random.uniform(-math.sqrt(6), math.sqrt(6), (n, p))
    elif distribution == "laplace":
        return np.random.laplace(0, 1, (n, p))
    elif distribution == "logistic":
        return np.random.logistic(0, math.sqrt(6)/math.pi, (n,p))
    else:
        return "Choose distribution in 'uniform', 'laplace', or 'logistic'"

In [ ]:
X1 = generator(1000, 200, "uniform")
X2 = generator(1000, 500, "uniform")
X3 = generator(1000, 1000, "uniform")
X4 = generator(1000, 2000, "uniform")

S1 = np.linalg.svd(X[1], compute_uv = False)
S2 = np.linalg.svd(X[2], compute_uv = False)
S3 = np.linalg.svd(X[3], compute_uv = False)
S4 = np.linalg.svd(X[4], compute_uv = False)

plt.plot(S1, label = "p=200")
plt.plot(S2, label = "p=500")
plt.plot(S3, label = "p=1000")
plt.plot(S4, label = "p=2000")
plt.xlabel('Singular value index')
plt.ylabel('Singular value value')
plt.title('Singular values of X, matrix of M1000,p, distributed uniformly with mean 0 and variance 2')
plt.legend()

In [ ]:
Y1 = generator(1000, 200, "laplace")
Y2 = generator(1000, 500, "laplace")
Y3 = generator(1000, 1000, "laplace")
Y4 = generator(1000, 2000, "laplace")

S1 = np.linalg.svd(Y1, compute_uv = False)
S2 = np.linalg.svd(Y2, compute_uv = False)
S3 = np.linalg.svd(Y3, compute_uv = False)
S4 = np.linalg.svd(Y4, compute_uv = False)

plt.plot(S1, label = "p=200")
plt.plot(S2, label = "p=500")
plt.plot(S3, label = "p=1000")
plt.plot(S4, label = "p=2000")
plt.xlabel('Singular value index')
plt.ylabel('Singular value value')
plt.title('Singular values of Y, matrix of M1000,p, distributed by a Laplace law with mean 0 and variance 2')
plt.legend()

In [ ]:
Z1 = generator(1000, 200, "logistic")
Z2 = generator(1000, 500, "logistic")
Z3 = generator(1000, 1000, "logistic")
Z4 = generator(1000, 2000, "logistic")

S1 = np.linalg.svd(Z1, compute_uv = False)
S2 = np.linalg.svd(Z2, compute_uv = False)
S3 = np.linalg.svd(Z3, compute_uv = False)
S4 = np.linalg.svd(Z4, compute_uv = False)

plt.plot(S1, label = "p=200")
plt.plot(S2, label = "p=500")
plt.plot(S3, label = "p=1000")
plt.plot(S4, label = "p=2000")
plt.xlabel('Singular value index')
plt.ylabel('Singular value value')
plt.title('Singular values of Z, matrix of M1000,p, distributed by a Logistic law with mean 0 and variance 2')
plt.legend()

5) Display on one single graph the spectrum (i.e. the set of eigen values) of XJX{n for
n “ 1000, and p “ 200, 500, 1000, 2000. Comment.

In [ ]:
n = 1000
X1_prime = X1.T @ X1 / n
X2_prime = X2.T @ X2 / n
X3_prime = X3.T @ X3 / n
X4_prime = X4.T @ X4 / n

# Getting the vectors of eigen values (sorted in desc order) 
W1 = LA.eig(X1_prime)[0]
W1[::-1].sort()
W2 = LA.eig(X2_prime)[0]
W2[::-1].sort()
W3 = LA.eig(X3_prime)[0]
W3[::-1].sort()
W4 = LA.eig(X4_prime)[0]
W4[::-1].sort()

plt.plot(W1, label = "p=200")
plt.plot(W2, label = "p=500")
plt.plot(W3, label = "p=1000")
plt.plot(W4, label = "p=2000")
plt.title("Eigenvalues of X.T@X/n of Mp(R), where X is distributed with a uniform of mean 0 and variance 2")
plt.xlabel("Eigenvalue index")
plt.ylabel("Eigenvalue value")
plt.xlim(0, 1000)
plt.legend()

In [ ]:
Y1_prime = Y1.T @ Y1 / n
Y2_prime = Y2.T @ Y2 / n
Y3_prime = Y3.T @ Y3 / n
Y4_prime = Y4.T @ Y4 / n

# Get the vectors of eigen values (sorted in desc order) W1, W2, W3, W4
W1 = LA.eig(Y1_prime)[0]
W1[::-1].sort()
W2 = LA.eig(Y2_prime)[0]
W2[::-1].sort()
W3 = LA.eig(Y3_prime)[0]
W3[::-1].sort()
W4 = LA.eig(Y4_prime)[0]
W4[::-1].sort()

plt.plot(W1, label = "p=200")
plt.plot(W2, label = "p=500")
plt.plot(W3, label = "p=1000")
plt.plot(W4, label = "p=2000")
plt.title("Eigenvalues of X.T@X/n of Mp(R), where X is distributed with a Laplace of mean 0 and variance 2")
plt.xlabel("Eigenvalue index")
plt.ylabel("Eigenvalue value")
plt.xlim(0, 1000)
plt.legend()

In [ ]:
Z1_prime = Z1.T @ Z1 / n
Z2_prime = Z2.T @ Z2 / n
Z3_prime = Z3.T @ Z3 / n
Z4_prime = Z4.T @ Z4 / n

# Get the vectors of eigen values (sorted in desc order) W1, W2, W3, W4
W1 = LA.eig(Z1_prime)[0]
W1[::-1].sort()
W2 = LA.eig(Z2_prime)[0]
W2[::-1].sort()
W3 = LA.eig(Z3_prime)[0]
W3[::-1].sort()
W4 = LA.eig(Z4_prime)[0]
W4[::-1].sort()
# Plot
plt.plot(W1, label = "p=200")
plt.plot(W2, label = "p=500")
plt.plot(W3, label = "p=1000")
plt.plot(W4, label = "p=2000")
plt.title("Eigenvalues of X.T@X/n of Mp(R), where X is distributed with a Logistic law of mean 0 and variance 2")
plt.xlabel("Eigenvalue index")
plt.ylabel("Eigenvalue value")
plt.xlim(0, 1000)
plt.legend()

Comments <br>
<br>
X matrix of M1000,p is distributed with a uniform probability and has mean 0 and var 2 <br>
Y matrix of M1000,p is distributed with a Laplace probability and has mean 0 and var 2 <br>
Z matrix of M1000,p is distributed with a logistic probability and has mean 0 and var 2 <br>
<br>
We create X1, X2, X3, X4 by fixing n=1000 and increasing p (from 200, 500, 1000 and 2000).<br>
We create Y1, Y2, Y3, Y4 by fixing n=1000 and increasing p (from 200, 500, 1000 and 2000).<br>
We create Z1, Z2, Z3, Z4 by fixing n=1000 and increasing p (from 200, 500, 1000 and 2000).<br>
<br>
In the first question, we plot the singular values of Xi, Yi, Zi for various values of p. <br>
I.E. we plot the sqrts of the eigen values of Xi.T @ Xi, Yi.T @ Yi, Zi.T @ Zi. <br>
For all 3 distributions, we get the same plot: <br>
-The # of non negative singular values = the # of columns. <br>
-The values of those singluar values have greater amplitude (higher maximum, lower minimum) as p increases <br>
-The singular values decrease in a linear fashion <br>
<br>
In the second question, we plot the eigenvalues of Xi.T @ Xi / n, Yi.T @ Yi / n, Zi.T @ Zi / n, where n=1000 <br>
I.E. we plot the eigevalues * 1/n of Xi.T @ Xi, Yi.T @ Yi, Zi.T @ Zi.  <br>
For all 3 distributions, we get the same plot.  <br>
-The # of non negative eigen values = the # of columns. <br>
-The values of those eigen values have greater maximum as p increases, and all tend to reach 0 <br>
-The singular values decrease in a quadratic fashion <br>

Exercice 3

6) Write a function coding the power Algorithm

In [ ]:
def power_method(X, n, u, v):
    for j in range(0, n):
        u = X @ v 
        v = X.T @ u
        v = v/LA.norm(v)
        u = u/LA.norm(u)
    return (u,v)

7) Illustrate visually (with a graph) the convergence of the algorithm iterates. Is it true that
the output u, v from the algorithm converge to the singular vector associated to the largest
singular value of X (check by computing these with numpy built in functions) ? Bonus :
can you show it mathematically ?

In [ ]:
# Initialization
n = 100
p = 2000
n_iter = 250

X = np.random.normal(loc=0, scale=5, size=(n, p)) 
u = np.random.rand(n, 1) 
v = np.random.rand(p, 1)

U, s, VT = svd(X)

In [ ]:
# Convergence of U

# We create an array iter_u containing U for n_iter
def iter_u(X, n, u, v):
    iter_u = []
    for j in range(0, n):
        u = X @ v 
        v = X.T @ u
        v = v/LA.norm(v)
        u = u/LA.norm(u)
        iter_u.append(u)
    return(np.absolute(iter_u))

iter_u = iter_u(X, n_iter, u, v)

# We create an array limit_u containing the limit vector, 
# i.e. the left singular vector, for n_iter
limit_u = U[:,0].reshape(n, 1)
limit_u = np.full(shape=(n_iter, n, 1), fill_value=limit_u)
limit_u = np.absolute(limit_u)

# We substract the abs value of iter_u and the abs value of limit_u
# and we compute the L2-norm
conv_u = iter_u - limit_u
norm_conv_u = LA.norm(conv_u, axis=1)

# We plot n_iter and the L2-norm
plt.plot(norm_conv_u)
plt.title("Convergence of u to the left-singular vector of X")
plt.xlabel("Number of iterations of the power algorithm")
plt.ylabel("L2-Norm of u - u_limit")

In [ ]:
# Convergence of V

# We create an array iter_v containing V for n_iter
def iter_v(X, n, u, v):
    iter_v = []
    for j in range(0, n):
        u = X @ v 
        v = X.T @ u
        v = v/LA.norm(v)
        u = u/LA.norm(u)
        iter_v.append(v)
    return(np.absolute(iter_v))

iter_v = iter_v(X, n_iter, u, v)

# We create an array limit_v containing the limit vector, 
# i.e. the right singular vector, for n_iter 
limit_v = VT.transpose()[:,0].reshape(p,1)
limit_v = np.full(shape=(n_iter, p,1), fill_value=limit_v)
limit_v = np.absolute(limit_v)

# We substract the abs value of iter_v and the abs value of limit_v
# and we compute the L2-norm
conv_v = iter_v - limit_v
norm_conv_v = LA.norm(conv_v, axis=1)

# # We plot n_iter and the L2-norm 
plt.plot(norm_conv_v)
plt.title("Convergence of v to the right-singular vector of X")
plt.xlabel("Number of iterations of the power algorithm")
plt.ylabel("L2-Norm of v - v_limit")

8) Provide two sets of initialization vectors leading to different limits for this algorithm ;
explain how they are related.

In [ ]:
# We tried to choose u, v orthogonal to the limit vectors but the convergence was still valid

9) Provide a way to approximate the largest singular value of X using the power method.

We have generated a matrix X using a normal distribution <br>
The power method approximates ui and vi, the left and right singular vectors of the largest singular value of X <br>
We know that, if ui and vi are the left-singular vector and the right-singular vector associated with σi then, Avi=σiui and ATui=σivi.

In [ ]:
# Initialization

n = 100
p = 2000
n_iter = 300 # The # of iterations is important to get a good approximation

X = np.random.normal(loc=0, scale=5, size=(n, p)) 
u = np.random.rand(n, 1) 
v = np.random.rand(p, 1) 

u, v = power_method(X, n_iter, u, v)

In [ ]:
# First way to approximate 
sigma = (X @ v) / u
print(sigma[0])

# Second way to approximate 
sigma = (X.T @ u) / v
print(sigma[0])

In [ ]:
# Getting the actual value of sigma through svd
U, s, VT = svd(X)
print(s[0])

In [ ]:
# Comparing approximation and actual value
np.isclose(sigma[0], s[0]) 

NB. On note que le nombre d'itérations n_iter doit être suffisamment élevé pour obtenir une approximation correcte
Pour ce faire, on détermine n_iter en utilisant les représentations graphiques obtenu à la question 7. <br>
Pour améliorer cet algorithme on chercherait à déterminer n automatiquement

10) Bonus : build upon the power method to provide an algorithm that can approximate the
second largest singular value of X (without using an SVD function)

Exercice 4. (PCA)

11) Import with Pandas the dataset defra_consumption.csv

In [ ]:
df = pd.read_csv('http://josephsalmon.eu/enseignement/TELECOM/SD204/defra_consumption.csv', sep=';')
df.rename({'Unnamed: 0': 'Category'}, axis=1, inplace=True)
df

12) Center and standardize the dataset using the preprocessing module from sklearn, where
we write X P Rnˆp for the associated matrix (n “ 17 is the number of observations, and p “ 4 is the number of variables)

In [ ]:
num_cols = [c for c in df.columns
                     if df[c].dtype.kind in ["i", "f"]]

df[num_cols] = preprocessing.scale(df[num_cols]) 

X = df.iloc[:,1:].to_numpy()

13) Display a scatter plot of the n points projected on the space generated by the first two
principal axes (corresponding to largest eigenvalues).

Visualizing 2 or 3 dimensional data is not that challenging. 
However, our dataset is 4 dimensional. 
We will use PCA to reduce that 4 dimensional data into 2 or 3 dimensions so that we can plot and hopefully understand the data better. <br>
In this section, the code projects the original data which is 4 dimensional into 2 dimensions. We note that after dimensionality reduction, there usually isn’t a particular meaning assigned to each principal component. The new components are just the two main dimensions of variation.

In [ ]:
from sklearn.decomposition import PCA
pca = PCA(n_components=2)
principalComponents = pca.fit_transform(X)

pca_df = pd.DataFrame(data = principalComponents
             , columns = ['PC1', 'PC2'])
pca_df = pd.concat([pca_df, df[['Category']]], axis = 1)

plt.figure(figsize=(10,6))
sns.scatterplot(data=pca_df, x="PC1", y="PC2", s=100)
plt.xlabel("PC1", fontsize = "20")
plt.ylabel("PC2", fontsize = "20")

# Add labels : try using Adjusttext

The explained variance tells how much information (variance) can be attributed to each of the principal components. <br>
Here, we notice that, although the 2 compoents contain 99% of information together, the first component contains much more information that the second one (97% vs 2%)

In [ ]:
pca.explained_variance_ratio_

14) Repeat the previous question for the space generated by three axes.

In [ ]:
pca = PCA(n_components=3)
principalComponents = pca.fit_transform(X)
pca_df_2 = pd.DataFrame(data = principalComponents
             , columns = ['PC1', 'PC2', 'PC3'])

fig = plt.figure(figsize=(20,10))
ax = fig.add_subplot(111, projection='3d')
ax.scatter(pca_df_2['PC1'], pca_df_2['PC2'], pca_df_2['PC3'], s=70)

plt.title("3 dimensions")
ax.view_init(15, 260)
plt.show()

In [ ]:
pca.explained_variance_ratio_

15) Compare the previous 2D and 3D graphs with the one obtained as follows :

(a) Compute X.T @ X, and diagonalize it. Project the n points encoded by X over the span
of the eigen vectors associated to the two (respectively three) largest eigen values.

In [ ]:
# 2 dimensions

start11 = time.time()

eigen_values, eigen_vectors = LA.eig(X.T @ X)

idx = eigen_values.argsort()[::-1]   
eigen_values = eigen_values[idx]
eigen_vectors = eigen_vectors[:,idx]

P = np.array(eigen_vectors)
M = X @ P[:, :2]
pcomp = pd.DataFrame(data = M
             , columns = ['PC1', 'PC2'])

time11 = time.time() - start11

plt.figure(figsize=(10,6))
sns.scatterplot(data=pcomp, x="PC1", y="PC2", s=100)
plt.xlabel("PC1", fontsize = "20")
plt.ylabel("PC2", fontsize = "20")
sns.set(rc={'figure.figsize':(15,15)})

In [ ]:
# 3 dimensions

start12 = time.time()

eigen_values, eigen_vectors = LA.eig(X.T @ X)

idx = eigen_values.argsort()[::-1]   
eigen_values = eigen_values[idx]
eigen_vectors = eigen_vectors[:,idx]

P = np.array(eigen_vectors)
M = X @ P[:, :3]
pcomp = pd.DataFrame(data = M
                     , columns = ['PC1', 'PC2', 'PC3'])

time12 = time.time() - start12

fig = plt.figure(figsize=(20,10))
ax = fig.add_subplot(111, projection='3d')
ax.scatter(pcomp['PC1'], pcomp['PC2'], pcomp['PC3'], s=70)
plt.title("3 dimensions")
ax.view_init(15, 260)
plt.show()

In [ ]:
# AMELIORER CE GRAPH https://python-graph-gallery.com/372-3d-pca-result/

(b) Compute the SVD of X. Project the n points encoded by X P R
nˆp over the span of
the correct (right or left ?) singular vectors associated to the two (respectively three)
largest singular values 2
.

In [ ]:
start21 = time.time()

u, s, vh = np.linalg.svd(X, full_matrices=False)
V = vh.T

M = X @ V[:, :2]
pcomp = pd.DataFrame(data = M
             , columns = ['PC1', 'PC2'])

time21 = time.time() - start21

plt.figure(figsize=(10,6))
sns.scatterplot(data=pcomp, x="PC1", y="PC2", s=100)
plt.xlabel("PC1", fontsize = "20")
plt.ylabel("PC2", fontsize = "20")
sns.set(rc={'figure.figsize':(15,15)})

In [ ]:
start22 = time.time()

u, s, vh = np.linalg.svd(X, full_matrices=False)
V = vh.T

M = X @ V[:, :3]
pcomp = pd.DataFrame(data = M
                     , columns = ['PC1', 'PC2', 'PC3'])

time22 = time.time() - start22

fig = plt.figure(figsize=(20,10))
ax = fig.add_subplot(111, projection='3d')
ax.scatter(pcomp['PC1'], pcomp['PC2'], pcomp['PC3'], s=70)
plt.title("3 dimensions")
ax.view_init(15, 260)
plt.show()

(c) Evaluate the difference in timing for the two methods.

In [ ]:
print("For dimension = 2, the time elapsed with method 1 (eigenvectors of X.T @ X) is %.5f and with method 2 (right singular vectors of X) is %.5f." % (time11, time21))
print("For dimension = 3, the time elapsed with method 1 (eigenvectors of X.T @ X) is %.5f and with method 2 (right singular vectors of X) is %.5f." % (time12, time22))

Method 1 is slightly slower if the number of dimensions n=2, and slighly faster if n=3

Exercice 5

In [2]:
df = pd.read_csv('auto-mpg.data-original', header=None, delim_whitespace=True,
            names=['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
                   'acceleration', 'model year', 'origin', 'car name'])
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0        8.0         307.0       130.0  3504.0          12.0   
1    15.0        8.0         350.0       165.0  3693.0          11.5   
2    18.0        8.0         318.0       150.0  3436.0          11.0   
3    16.0        8.0         304.0       150.0  3433.0          12.0   
4    17.0        8.0         302.0       140.0  3449.0          10.5   
..    ...        ...           ...         ...     ...           ...   
401  27.0        4.0         140.0        86.0  2790.0          15.6   
402  44.0        4.0          97.0        52.0  2130.0          24.6   
403  32.0        4.0         135.0        84.0  2295.0          11.6   
404  28.0        4.0         120.0        79.0  2625.0          18.6   
405  31.0        4.0         119.0        82.0  2720.0          19.4   

     model year  origin                   car name  
0          70.0     1.0  chevrolet chevelle malibu  
1          70.0     1.0          buick skylark 320  
2          70.0     1.0         plymouth satellite  
3          70.0     1.0              amc rebel sst  
4          70.0     1.0                ford torino  
..          ...     ...                        ...  
401        82.0     1.0            ford mustang gl  
402        82.0     2.0                  vw pickup  
403        82.0     1.0              dodge rampage  
404        82.0     1.0                ford ranger  
405        82.0     1.0                 chevy s-10  

[406 rows x 9 columns]

In [3]:
df.dtypes

mpg             float64
cylinders       float64
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
model year      float64
origin          float64
car name         object
dtype: object

Since all columns of interest have dtype float64, we can confidently say that there is no exotic string encoding for NaNs. <br>
NaNs could therefore be exotically encoded with floats (e.g. 999.99), so we use df.describe to spot "outliers", but cannot see any irrational values (e.g. negative values) <br>
Therefore, we should identify NaNs with the traditional df.isna() <br>

In [4]:
df.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  406.000000    406.000000  400.000000   406.000000   
mean    23.514573    5.475369    194.779557  105.082500  2979.413793   
std      7.815984    1.712160    104.922458   38.768779   847.004328   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    105.000000   75.750000  2226.500000   
50%     23.000000    4.000000    151.000000   95.000000  2822.500000   
75%     29.000000    8.000000    302.000000  130.000000  3618.250000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model year      origin  
count    406.000000  406.000000  406.000000  
mean      15.519704   75.921182    1.568966  
std        2.803359    3.748737    0.797479  
min        8.000000   70.000000    1.000000  
25%       13.700000   73.000000    1.000000  
50%       15.500000   76.000000    1.000000  
75%       17.175000   79.000000    2.000000  
max       24.800000   82.000000    3.000000

In [5]:
df.isna().sum()

mpg             8
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

In [6]:
df.dropna(inplace=True)

In [7]:
 df.isna().sum()

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

In [8]:
df.drop('car name', axis=1, inplace=True)

In [9]:
df.reset_index(drop=True, inplace=True)
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0        8.0         307.0       130.0  3504.0          12.0   
1    15.0        8.0         350.0       165.0  3693.0          11.5   
2    18.0        8.0         318.0       150.0  3436.0          11.0   
3    16.0        8.0         304.0       150.0  3433.0          12.0   
4    17.0        8.0         302.0       140.0  3449.0          10.5   
..    ...        ...           ...         ...     ...           ...   
387  27.0        4.0         140.0        86.0  2790.0          15.6   
388  44.0        4.0          97.0        52.0  2130.0          24.6   
389  32.0        4.0         135.0        84.0  2295.0          11.6   
390  28.0        4.0         120.0        79.0  2625.0          18.6   
391  31.0        4.0         119.0        82.0  2720.0          19.4   

     model year  origin  
0          70.0     1.0  
1          70.0     1.0  
2          70.0     1.0  
3          70.0     1.0  
4          70.0     1.0  
..          ...     ...  
387        82.0     1.0  
388        82.0     2.0  
389        82.0     1.0  
390        82.0     1.0  
391        82.0     1.0  

[392 rows x 8 columns]

17) Add two or three binary features to meaningfully encode the three origins ('origin' feature, for which, initially, 1 stands for USA, 2 for Europe and 3 for Japan) 3
.

Two methods are possible. 1) pd.get_dummies 2) Scikit Learn OneHotEncoder

In [10]:
# Before employing either method, we add the info
df['origin'].replace([1,2,3],['USA','Europe', 'Japan'],inplace=True)
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0        8.0         307.0       130.0  3504.0          12.0   
1    15.0        8.0         350.0       165.0  3693.0          11.5   
2    18.0        8.0         318.0       150.0  3436.0          11.0   
3    16.0        8.0         304.0       150.0  3433.0          12.0   
4    17.0        8.0         302.0       140.0  3449.0          10.5   
..    ...        ...           ...         ...     ...           ...   
387  27.0        4.0         140.0        86.0  2790.0          15.6   
388  44.0        4.0          97.0        52.0  2130.0          24.6   
389  32.0        4.0         135.0        84.0  2295.0          11.6   
390  28.0        4.0         120.0        79.0  2625.0          18.6   
391  31.0        4.0         119.0        82.0  2720.0          19.4   

     model year  origin  
0          70.0     USA  
1          70.0     USA  
2          70.0     USA  
3          70.0     USA  
4          70.0     USA  
..          ...     ...  
387        82.0     USA  
388        82.0  Europe  
389        82.0     USA  
390        82.0     USA  
391        82.0     USA  

[392 rows x 8 columns]

In [12]:
# 1st solution using pd.get_dummies
df = pd.concat([df,pd.get_dummies(df['origin'],prefix='country')]
               ,axis=1).drop(['origin'],axis=1)
# drop_first=True to drop one dim
df

KeyError: 'origin'

In [11]:
# 2nd solution using Sklearn OneHotEncoder
enc = OneHotEncoder(handle_unknown='ignore')
enc_df = pd.DataFrame(enc.fit_transform(df[['origin']]).toarray())
enc_df.columns = enc.get_feature_names(['origin'])
df = df.join(enc_df)
df.drop('origin', axis=1, inplace=True)
df
# Can we put all of that into a pipeline?

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0        8.0         307.0       130.0  3504.0          12.0   
1    15.0        8.0         350.0       165.0  3693.0          11.5   
2    18.0        8.0         318.0       150.0  3436.0          11.0   
3    16.0        8.0         304.0       150.0  3433.0          12.0   
4    17.0        8.0         302.0       140.0  3449.0          10.5   
..    ...        ...           ...         ...     ...           ...   
387  27.0        4.0         140.0        86.0  2790.0          15.6   
388  44.0        4.0          97.0        52.0  2130.0          24.6   
389  32.0        4.0         135.0        84.0  2295.0          11.6   
390  28.0        4.0         120.0        79.0  2625.0          18.6   
391  31.0        4.0         119.0        82.0  2720.0          19.4   

     model year  origin_Europe  origin_Japan  origin_USA  
0          70.0            0.0           0.0         1.0  
1          70.0            0.0           0.0         1.0  
2          70.0            0.0           0.0         1.0  
3          70.0            0.0           0.0         1.0  
4          70.0            0.0           0.0         1.0  
..          ...            ...           ...         ...  
387        82.0            0.0           0.0         1.0  
388        82.0            1.0           0.0         0.0  
389        82.0            0.0           0.0         1.0  
390        82.0            0.0           0.0         1.0  
391        82.0            0.0           0.0         1.0  

[392 rows x 10 columns]

NB. We tried to add the OneHotEncoding to the pipeline (to use it in the last question), but we cannot easily rename the columns and drop the original "origin" column

18) Select (manually) 9 rows of the dataset such that all 3 origins are represented, and model
year is not constant. Get the least-squares estimator θˆ (with intercept) the prediction vector
yˆ, considering only these 9 lines. What do you observe ? Why ?

Should we add part of the pipeline here?

In [ ]:
df2 = df.drop_duplicates('model year').sample(9).reset_index(drop=True)
df2

In [ ]:
# Multiple linear regression with Scikit learn
X = df2.drop('mpg',axis=1)
y = df2.mpg

reg = LinearRegression().fit(X, y)

theta_hat = reg.coef_
intercept = reg.intercept_
y_hat = X @ theta_hat + intercept 
y_hat

print(theta_hat, intercept, y_hat, reg.score(X,y))



Each time we change the sample, we see that the weights are different - we even
sometimes get illogical weights (e.g. a positive weight for the "weight" feature, which 
means that the heavier the car, the higher the mpg, (and the less the car "consumes" fuel)) <br>
We can see that the r2 score is 1 (y_hat = y), wich might show that we are "overfitting"
the training set <br>
Therefore, we cannot safely extrapolate those weights to new
data because our training set is too small <br>

19) Now, get the least-squares estimator θˆ and the prediction vector yˆ (with intercept) over the
whole dataset, after performing scaling/centering (the columns must have unit standard
deviation and be zero mean). Which variables seem to best explain gasoline consumption
according to your model ? Why wouldn’t this answer make sense if the columns were not
normalized ?

In [ ]:
df

In [ ]:
X = df.drop('mpg',axis=1).values
y = df.mpg  
print(X, y)

In [ ]:
pipeline = Pipeline(steps=[('scaler', StandardScaler(copy=False)), ('regression', LinearRegression())])
pipeline.fit(X, y)
reg = pipeline.named_steps['regression']
theta_hat = reg.coef_
intercept = reg.intercept_
print(theta_hat, intercept)

In [ ]:
X

We get r2 = 0.82, i.e. 82% of the var of the mpg can be explained by our weights.
Weight is the most important feature. If the columns were not normalized, this would
not make sense because the scale of weight is much higher than the other features
and we would suspect the influence of weight to be overestimated.
Here, having scaled the features, we can confidently say that weight is an
important features, intependently of the range of values it takes

In [ ]:
y_hat = X @ theta_hat + intercept 
y_hat

20)

In [ ]:
r = y - y_hat
r_sq_norm = LA.norm(r)**2
r_sq_norm

In [ ]:
yn_bar = np.mean(y)
yn_bar_vect = yn_bar * np.ones(len(y))
yn_bar_vect

In [ ]:
one = LA.norm(y - yn_bar_vect)**2
two = r_sq_norm + LA.norm(y_hat - yn_bar_vect)**2
print(one, two)
np.isclose(one, two)

21)

In [ ]:
X_test = pd.DataFrame(np.array([[6, 225, 100, 3233, 15.4, 117, 0, 0, 1]]), 
                      columns=['cylinders', 'displacement', 'horsepower', 'weight',
                   'acceleration', 'model year', 'country_Europe', 'country_Japan', 'country_USA'])
X_test

In [ ]:
pipeline.predict(X_test)

In [ ]:
# Double checking the result
y_hat = X_test.values @ theta_hat + intercept
y_hat

In [ ]:
df.describe()

53.1 MPG seems very high, indeed higher than the maximum (46.6), which means that the consumption is very low. <br>
In our original model, the model year have a weight of 2.86, which is the highest weight, which means that the more recent the car, the higher the mpg (and the less it consumes fuel) <br>
Here, we get a very high mpg because we are extrapolating the weights with model year 117, which is much bigger than what we had in our training data. <br>
It would be interesting to make further analysis to understand if this linear relation is still true 35 years after the training data (it could have been exponential due to hybrid cars for instance?)